src/hardware/adc.rs

```diff
@@ -1,22 +1,44 @@
 // Thermostat ADC struct.
 
+use defmt::Format;
 use num_enum::TryFromPrimitive;
-use shared_bus_rtic::SharedBus;
 
 use super::ad7172;
 
 use super::hal::{
-    gpio::{gpioe::*, Alternate, Output, PushPull, AF5},
+    gpio::{
+        gpiob::*, gpioc::*, gpioe::*, Alternate, ExtiPin, Input, Output, PullUp, PushPull, AF5,
+    },
     hal::blocking::delay::DelayUs,
     hal::digital::v2::OutputPin,
+    hal::digital::v2::PinState::{High, Low},
     prelude::*,
     rcc::{rec, CoreClocks},
     spi,
     spi::{Enabled, Spi},
     stm32::SPI4,
 };
 
-#[derive(Clone, Copy, TryFromPrimitive)]
+macro_rules! set_cs {
+    ($self:ident, $phy:ident, $state:ident) => {
+        match $phy {
+            AdcPhy::Zero => $self.cs.0.set_state($state).unwrap(),
+            AdcPhy::One => $self.cs.1.set_state($state).unwrap(),
+            AdcPhy::Two => $self.cs.2.set_state($state).unwrap(),
+            AdcPhy::Three => $self.cs.3.set_state($state).unwrap(),
+        };
+    };
+}
+
+macro_rules! setup_adc {
```

Comment:
Overall this macro saves you ~one line of code. I don't think it's very useful. It doesn't help me reading your code.